src/flint/test/test_all.py

```diff
@@ -2631,6 +2633,398 @@ def setbad(obj, i, val):
+        assert mpoly({(0, 0): 1, (0, 1): 1, (1, 0): 1}).degrees() == {"x0": 1, "x1": 1}
+        assert mpoly({(0, 0): 1, (0, 1): 1, (1, 0): 1, (2, 2): 2}).degrees() == {"x0": 2, "x1": 2}
+
+        assert (P(1, ctx=ctx) == P(1, ctx=ctx)) is True
+        assert (P(1, ctx=ctx) != P(1, ctx=ctx)) is False
+        assert (P(1, ctx=ctx) == P(2, ctx=ctx)) is False
+        assert (P(1, ctx=ctx) != P(2, ctx=ctx)) is True
+
+        assert (P(1, ctx=ctx) == 1) is False
+        assert (P(1, ctx=ctx) != 1) is True
+        assert (1 == P(1, ctx=ctx)) is False
+        assert (1 != P(1, ctx=ctx)) is True
+
+        assert (P(1, ctx=ctx) == P(1, ctx=ctx1)) is False
+        assert (P(1, ctx=ctx) != P(1, ctx=ctx1)) is True
+
+        assert (P(1, ctx=ctx) == None) is False
+        assert (P(1, ctx=ctx) != None) is True
+        assert (None == P(1, ctx=ctx)) is False
+        assert (None != P(1, ctx=ctx)) is True
+
+        assert P(ctx.from_dict({(0, 1): 3})) == ctx.from_dict({(0, 1): 3})
+        assert P({(0, 1): 3}, ctx=ctx) == ctx.from_dict({(0, 1): 3})
+
+        if P is flint.fmpq_mpoly:
+            ctx_z = flint.fmpz_mpoly_ctx.get_context(2)
+            assert quick_poly() == P(ctx_z.from_dict({(0, 0): 1, (0, 1): 2, (1, 0): 3, (2, 2): 4}))
+            assert P(ctx_z.from_dict({(0, 0): 1}), ctx=ctx) == P({(0, 0): 1}, ctx=ctx)
+
+        assert raises(lambda: P(ctx=ctx) < P(ctx=ctx), TypeError)
+        assert raises(lambda: P(ctx=ctx) <= P(ctx=ctx), TypeError)
+        assert raises(lambda: P(ctx=ctx) > P(ctx=ctx), TypeError)
+        assert raises(lambda: P(ctx=ctx) >= P(ctx=ctx), TypeError)
+        assert raises(lambda: P(ctx=ctx) < None, TypeError)
+        assert raises(lambda: P(ctx=ctx) <= None, TypeError)
+        assert raises(lambda: P(ctx=ctx) > None, TypeError)
+        assert raises(lambda: P(ctx=ctx) >= None, TypeError)
+        assert raises(lambda: None < P(ctx=ctx), TypeError)
+        assert raises(lambda: None <= P(ctx=ctx), TypeError)
+        assert raises(lambda: None > P(ctx=ctx), TypeError)
+        assert raises(lambda: None >= P(ctx=ctx), TypeError)
+
+        p = quick_poly()
+        assert p.coefficient(2) == S(2)
+        assert raises(lambda: p.coefficient(-1), IndexError)
+        assert raises(lambda: p.coefficient(10), IndexError)
+
+        assert p[0] == mpoly({(2, 2): 4})
+        assert p[3] == mpoly({(0, 0): 1})
+        assert raises(lambda: p[-1], IndexError)
+        assert raises(lambda: p[4], IndexError)
+
+        assert p[(2, 2)] == 4
+        assert p[(0, 0)] == 1
+        assert raises(lambda: p[(1,)], ValueError)
+        assert raises(lambda: p[(1, "bad")], TypeError)
+        assert raises(lambda: p["bad"], TypeError)
+
+        p = quick_poly()
+        p[1] = S(10)
+        assert p == mpoly({(0, 0): 1, (0, 1): 2, (1, 0): 10, (2, 2): 4})
+
+        p = quick_poly()
+        p[(1, 0)] = S(10)
+        assert p == mpoly({(0, 0): 1, (0, 1): 2, (1, 0): 10, (2, 2): 4})
```

Comment:
I find it a bit odd that `__getitem__` mixes sequence and dict-like behaviour.

Is there much use for being able to index the polynomial with integer indices like this?

Reply:
Nothing more than I thought that behaviour would be useful. With respect to your other comments, I'll make the behaviour consistent with univariate polynomials, and move the tuple index into a separate `get` method. That will mean indexing returns the coefficient of the `i`th term.

Reply:
I don't think that it can be made consistent with univariate polynomials. The `i`th term cannot be defined in a consistent way for multivariate polynomials. For univariate polynomials it can refer to terms not present in the polynomial:
```python
In [15]: p = fmpz_poly([1, 2])

In [16]: p
Out[16]: 2*x + 1

In [17]: p[5]
Out[17]: 0

In [18]: p[5] = 2

In [19]: p
Out[19]: 2*x^5 + 2*x + 1

In [20]: len(p)
Out[20]: 6

In [21]: list(p)
Out[21]: [1, 2, 0, 0, 0, 2]
```
While multivariate polynomials necessarily need to be different we should ensure that if they have sequence like methods they do consistently behave like a sequence.

I would probably go the other way and make them dict-like with tuple keys although I haven't looked closely at the access methods and how they are implemented in Flint to see what makes more sense.

Reply:
Flint has exposed functions that make both viable